components/collator/src/comparison.rs

```diff
@@ -1543,4 +1631,808 @@ impl CollatorBorrowed<'_> {
+    /// Write a single byte into the writer.
+    fn write_byte(&mut self, b: u8) -> Result<(), core::fmt::Error>;
+
+    /// Write leading bytes up to a zero byte, but always write at least one byte.
+    fn write_to_zero(&mut self, buf: &[u8]) -> Result<(), core::fmt::Error>;
+}
+
+impl<T> CollationKeySinkExt for T
+where
+    T: CollationKeySink,
+{
+    fn write_byte(&mut self, b: u8) -> Result<(), core::fmt::Error> {
+        self.write(&[b])?;
+        Ok(())
+    }
+
+    // guaranteed to get at least one byte and the rest are checked incrementally
+    #[allow(clippy::indexing_slicing)]
+    fn write_to_zero(&mut self, buf: &[u8]) -> Result<(), core::fmt::Error> {
+        self.write_byte(buf[0])?;
+        if buf.len() > 1 && buf[1] != 0 {
+            self.write_byte(buf[1])?;
+            if buf.len() > 2 && buf[2] != 0 {
+                self.write_byte(buf[2])?;
+            }
+        }
+        Ok(())
+    }
+}
+
+#[derive(Default)]
+struct SortKeyLevel {
+    buf: SmallVec<[u8; 40]>,
+}
+
+impl SortKeyLevel {
+    fn len(&self) -> usize {
+        self.buf.len()
+    }
+
+    fn is_empty(&self) -> bool {
+        self.buf.is_empty()
+    }
+
+    fn append_byte(&mut self, x: u8) {
+        self.buf.push(x);
+    }
+
+    fn append_weight_16(&mut self, w: u16) {
+        debug_assert_ne!(w, 0);
+        let b0 = (w >> 8) as u8;
+        let b1 = w as u8;
+        self.append_byte(b0);
+        if b1 != 0 {
+            self.append_byte(b1);
+        }
+    }
+
+    fn append_reverse_weight_16(&mut self, w: u16) {
+        debug_assert_ne!(w, 0);
+        let b0 = (w >> 8) as u8;
+        let b1 = w as u8;
+        if b1 != 0 {
+            self.append_byte(b1);
+        }
+        self.append_byte(b0);
+    }
+
+    fn append_weight_32(&mut self, w: u32) {
+        debug_assert_ne!(w, 0);
+        let b0 = (w >> 24) as u8;
+        let b1 = (w >> 16) as u8;
+        let b2 = (w >> 8) as u8;
+        let b3 = w as u8;
+        self.append_byte(b0);
+        if b1 != 0 {
+            self.append_byte(b1);
+            if b2 != 0 {
+                self.append_byte(b2);
+                if b3 != 0 {
+                    self.append_byte(b3);
+                }
+            }
+        }
+    }
+}
+
+impl AsRef<SmallVec<[u8; 40]>> for SortKeyLevel {
+    fn as_ref(&self) -> &SmallVec<[u8; 40]> {
+        &self.buf
+    }
+}
+
+impl AsMut<SmallVec<[u8; 40]>> for SortKeyLevel {
+    fn as_mut(&mut self) -> &mut SmallVec<[u8; 40]> {
+        &mut self.buf
+    }
+}
+
+struct SinkAdapter<'a, W> {
+    inner: &'a mut W,
+}
+
+impl<'a, W> SinkAdapter<'a, W> {
+    fn new(inner: &'a mut W) -> Self {
+        Self { inner }
+    }
+}
+
+impl<W> core::fmt::Write for SinkAdapter<'_, W>
+where
+    W: CollationKeySink,
+{
+    fn write_str(&mut self, s: &str) -> core::fmt::Result {
+        self.inner.write(s.as_bytes())?;
+        Ok(())
+    }
+}
+
+impl<W> write16::Write16 for SinkAdapter<'_, W>
+where
+    W: CollationKeySink,
+{
```

Comment:
As a micro optimization, please provide an implementation like this to encode single chars directly to UTF-8:

```rust
    fn write_char(&mut self, c: char) -> core::fmt::Result {
        let mut buf = [0u8; 4];
        self.inner.write(c.encode_utf8(&mut buf).as_bytes())
    }
```